pkg/core/resources/apis/hostnamegenerator/hostname/generator.go

```diff
@@ -0,0 +1,224 @@
+package hostname
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	kube_meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	hostnamegenerator_api "github.com/kumahq/kuma/pkg/core/resources/apis/hostnamegenerator/api/v1alpha1"
+	meshservice_api "github.com/kumahq/kuma/pkg/core/resources/apis/meshservice/api/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/core/runtime/component"
+	"github.com/kumahq/kuma/pkg/core/user"
+	core_metrics "github.com/kumahq/kuma/pkg/metrics"
+	"github.com/kumahq/kuma/pkg/util/maps"
+)
+
+type Generator struct {
+	logger     logr.Logger
+	interval   time.Duration
+	metric     prometheus.Summary
+	resManager manager.ResourceManager
+}
+
+var _ component.Component = &Generator{}
+
+func NewGenerator(
+	logger logr.Logger,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	interval time.Duration,
+) (*Generator, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_hostname_generator_ms",
+		Help:       "Summary of hostname generator interval",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+
+	return &Generator{
+		logger:     logger,
+		resManager: resManager,
+		interval:   interval,
+		metric:     metric,
+	}, nil
+}
+
+func (g *Generator) Start(stop <-chan struct{}) error {
+	g.logger.Info("starting")
+	ticker := time.NewTicker(g.interval)
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	for {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			if err := g.generateHostnames(ctx); err != nil {
+				return err
+			}
+			g.metric.Observe(float64(time.Since(start).Milliseconds()))
+		case <-stop:
+			g.logger.Info("stopping")
+			return nil
+		}
+	}
+}
+
+func apply(generator *hostnamegenerator_api.HostnameGeneratorResource, service *meshservice_api.MeshServiceResource) (string, error) {
+	if !generator.Spec.Selector.MeshService.Matches(service.Meta.GetLabels()) {
+		return "", nil
+	}
+	sb := strings.Builder{}
+	tmpl := template.New("").Funcs(
+		map[string]any{
+			"label": func(key string) (string, error) {
+				val, ok := service.GetMeta().GetLabels()[key]
+				if !ok {
+					return "", errors.Errorf("label %s not found", key)
+				}
+				return val, nil
+			},
+		},
+	)
+	tmpl, err := tmpl.Parse(generator.Spec.Template)
```

Comment:
Do we validate this statically on resource apply as well?

Reply:
added